### PR TITLE
LMB-519 | Process deltas mandatees decisions endpoint

### PR DESCRIPTION
**Pull request by JonasVanHoof**
## Description
Previously we where handing the differences and the process of those in the dispatch configuration of the consumer. In this PR we moved the logic from the dispatch config to an endpoint in this service. This makes it easier to process the data and it will not block jobs from executing.

[x] Log the differences between mandatees
[x] Process the differences of these mandatees

## How to test

1. Run the mandatees-decisions-consumer-app; LMB app; mandataris service with debug compose
2. The besluiten app should have created delta files in `/data/files/deltas/mandatees-decisons/<all the dates>`
3. When there is information in there the deltas will be catched by the mandataris service consumer passed on to the endpoint
4. When the endpoint is called all information should be processed correctly

## Links to other PR's

- Mandatees-decisions-app | https://github.com/JonasVanHoof/mandaten-besluit-consumer-app/pull/6 
- LMB-app |https://github.com/lblod/app-lokaal-mandatenbeheer/pull/163

### CODE REVIEW

File: util/types.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+export type Term = {
```
Comment (JonasVanHoof) [inline util/types.ts:1]: Do we want to keep the types.ts in the util folder or in the root?

File: docker-compose.debug.yml [hunk 2]
```diff
@@ -20,6 +20,35 @@ services:
...
+      DCR_SYNC_BASE_URL: 'http://producer-identifier'
+      DCR_JOB_CREATOR_URI: 'http://lblod.data.gift/services/delta-producer-background-jobs-initiator-mandatees-decisions'
+      DCR_SYNC_DATASET_SUBJECT: 'http://data.lblod.info/datasets/delta-producer/dumps/Mandatees-DecisionsCacheGraphDump'
+      DCR_INITIAL_SYNC_JOB_OPERATION: 'http://redpencil.data.gift/id/jobs/concept/JobOperation/deltas/initialPublicationGraphSyncing/mandatees-decisions'
+      DCR_DELTA_SYNC_JOB_OPERATION: 'http://redpencil.data.gift/id/jobs/concept/JobOperation/deltas/deltaDumpFileCreation/mandatees-decisions'
+      DCR_START_FROM_DELTA_TIMESTAMP: '2024-01-12T15:51:23.478Z'
+      DCR_SYNC_FILES_PATH: '/sync/mandatees-decisions/files'
+      DCR_KEEP_DELTA_FILES: true
+      DCR_DELTA_FILE_FOLDER: '/deltas/mandatees-decisions/files'
+      DCR_CRON_PATTERN_DELTA_SYNC: '0 * * * *	' # Every hour
```
Comment (JonasVanHoof) [inline docker-compose.debug.yml:39]: Update this in your override file to be  `DCR_CRON_PATTERN_DELTA_SYNC: '0/15 * * * * *' # Every 15 seconds minute`

File: config/custom-dispatching/util/batch-update.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
+async function parallelisedBatchedUpdate(
```
Comment (JonasVanHoof) [inline config/custom-dispatching/util/batch-update.js:1]: Content of this file is copied over from the `mandataris-besluiten-consumer-app`

File: config/custom-dispatching/initial-sync-dispatching.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+ *              graph: "<http://foo>",
+ *              subject: "<http://bar>",
+ *              predicate: "<http://baz>",
+ *              object: "<http://boom>^^<http://datatype>"
+ *            }
+ *         ]
+ * @return {void} Nothing
+ */
+async function dispatch(lib, data) {
+  console.log(`|> INITAL SYNC`);
```
Comment (JonasVanHoof) [inline config/custom-dispatching/initial-sync-dispatching.js:26]: Todo: remove

File: config/custom-dispatching/delta-sync-dispatching.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ *              graph: "<http://foo>",
+ *              subject: "<http://bar>",
+ *              predicate: "<http://baz>",
+ *              object: "<http://boom>^^<http://datatype>"
+ *            }
+ *         ]
+ * @return {void} Nothing
+ */
+async function dispatch(lib, data) {
+  console.log(`|> DELTA SYNC`);
```
Comment (JonasVanHoof) [inline config/custom-dispatching/delta-sync-dispatching.js:25]: Todo: remove

File: app.ts [hunk 6]
```diff
@@ -1,15 +1,28 @@
...
 import { mandatarissenRouter } from './routes/mandatarissen';
-import express, { ErrorRequestHandler } from 'express';
+import express, { Request, ErrorRequestHandler } from 'express';
+import bodyParser from 'body-parser';
+
 import { burgemeesterRouter } from './routes/burgemeester-benoeming';
 import { installatievergaderingRouter } from './routes/intallatievergadering';
+import { mandateesDecisionsRouter } from './routes/mandatees-decisions';
+
+app.use(
```
Comment (JonasVanHoof) [inline app.ts:10]: This is needed so we can get the information (`req.body`) passed by the delta notifier callback

**Review by JonasVanHoof — COMMENTED** [review]
WIP

**Comment by JonasVanHoof** [comment]
Updated some of the logs and fixed an issue with terminating a mandataris and got this result. Actually really happy that it is doing all of this!

```bash
mandataris-1  | |> 3.2 Mandataris exists in LMB database? false
mandataris-1  | |> 4 Persoon from mandataris: undefined.
mandataris-1  | |> 5 Mandaat for mandataris null
mandataris-1  | |> 5.2 No mandaat found for mandataris with subject: http://data.lblod.info/id/mandatarissen/02e55ec89ea809c499d478e557a1fa0c27a026bb1bc4357c0ea8777414cded37
mandataris-1  | |> 	 3.1 Start new loop http://data.lblod.info/id/mandatarissen/0c073784f7e3ff330a32121266e18d0f88171474d7f72ba683344ad868a9f6a5
mandataris-1  | |> 3.2 Mandataris exists in LMB database? false
mandataris-1  | |> 4 Persoon from mandataris: http://data.lblod.info/id/personen/170fa156c18ea6b1d90698326271de591344a8f63bc38cd630539c60d45d74a3.
mandataris-1  | |> 5 Mandaat for mandataris {
mandataris-1  | |> 6.1 persoon has overlapping mandaat? http://data.lblod.info/id/mandatarissen/0c073784f7e3ff330a32121266e18d0f88171474d7f72ba683344ad868a9f6a5
mandataris-1  | |> 6.2 Found start date for incoming mandataris? Tue Jan 01 2019 00:00:00 GMT+0000 (Coordinated Universal Time)
mandataris-1  | |> Terminated mandataris with uri: http://data.lblod.info/id/mandatarissen/0c073784f7e3ff330a32121266e18d0f88171474d7f72ba683344ad868a9f6a5.
mandataris-1  | |> 6.3 mandataris graph: http://mu.semte.ch/graph/organizations/d7647bc0a8e666de25cd429da54fc9e9909f886a177cb53df304f1273636f44d/LoketLB-mandaatGebruiker.
mandataris-1  | |> 6.4 inserting incoming triples
mandataris-1  | |> Inserted new mandataris quads (1) in graph (http://mu.semte.ch/graph/organizations/d7647bc0a8e666de25cd429da54fc9e9909f886a177cb53df304f1273636f44d/LoketLB-mandaatGebruiker).
mandataris-1  | |> End of logic for mandataris subject: http://data.lblod.info/id/mandatarissen/0c073784f7e3ff330a32121266e18d0f88171474d7f72ba683344ad868a9f6a5 
```

File: data-access/mandatees-decisions.ts [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+          DELETE {
+            GRAPH ?graph {
+              ${subjectPredicate} ${escaped.currentObject}
+            }
+          } INSERT {
+            GRAPH ${escaped.graph} {
+              ${subjectPredicate} ${escaped.incomingObject} .
+            }
+          } WHERE {
+             GRAPH ?graph {
```
Comment (JonasVanHoof) [inline data-access/mandatees-decisions.ts:149]: Use escaped.graph in query `GRAPH ${escaped.graph} {`

File: data-access/mandatees-decisions.ts [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+      } catch (error) {
+        throw Error(
+          `Could not insert incoming triple: ${JSON.stringify(incomingQuad)}`,
+        );
+      }
+    }
+  }
+}
+
+export async function findGraphOfType(rdfType: Term): Promise<Term> {
```
Comment (JonasVanHoof) [inline data-access/mandatees-decisions.ts:197]: check used?

File: util/sparql-result.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
+import { Quad, SparqlResult } from './types';
+
+export function getSparqlResults(sparqlResult: SparqlResult): Array<Quad> {
+  return sparqlResult.results.bindings;
+}
+
+export function findFirstSparqlResult(sparqlResult: SparqlResult): Quad | null {
```
Comment (JonasVanHoof) [inline util/sparql-result.ts:7]: type R = {
  [key: string]: Term
}

File: util/types.ts (hunk 1 above)
Comment (JonasVanHoof) [reply util/types.ts:1]: use the types.ts in root path

File: data-access/mandatees-decisions.ts [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+}
+
+export async function findPersoonForMandataris(
+  subject: Term,
+): Promise<null | Term> {
+  const queryForPersoon = `
+    PREFIX mandaat: <http://data.vlaanderen.be/ns/mandaat#>
+
+    SELECT ?object
+    WHERE {
```
Comment (JonasVanHoof) [inline data-access/mandatees-decisions.ts:99]: Minus history graqh

File: controllers/mandatees-decisions.ts [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+    );
+    if (isExistingInTarget) {
+      const currentQuads = await getValuesForSubjectPredicateInTarget(
+        incomingQuadsForSubject,
+      );
+      console.log('|> 3.3 Updating mandataris predicate values.');
+      await updateDifferencesOfMandataris(
+        currentQuads,
+        incomingQuadsForSubject,
+      );
```
Comment (JonasVanHoof) [inline controllers/mandatees-decisions.ts:49]: Early escape => `continue;`

File: controllers/mandatees-decisions.ts [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    // Looking for persoon in every graph!
+    const persoonOfMandataris =
```
Comment (JonasVanHoof) [inline controllers/mandatees-decisions.ts:53]: 1. check mandaat
2. Get graphe
3. get person from graph got from above logic

File: data-access/mandatees-decisions.ts (hunk 8 above)
Comment (JonasVanHoof) [reply data-access/mandatees-decisions.ts:197]: This is not used anymore

File: data-access/mandatees-decisions.ts [hunk 13]
```diff
@@ -0,0 +1,282 @@
...
+): Promise<null | Term> {
+  const escaped = {
+    graph: sparqlEscapeTermValue(graph),
+    mandataris: sparqlEscapeTermValue(subject),
+  };
+
+  const queryForPersoon = `
+    PREFIX mandaat: <http://data.vlaanderen.be/ns/mandaat#>
+
+    SELECT ?object
```
Comment (JonasVanHoof) [inline data-access/mandatees-decisions.ts:104]: Should i change the values of the select to the values that are get? The object was used as we expected return type Quad but that changed now. We could do `select  ?persoon`

Comment (bfidlers) [reply data-access/mandatees-decisions.ts:104]: yeah if you use ?person that might be a bit more clear.

**Review by bfidlers — APPROVED** [review]
We might need to refactor this a bit over time, to cut down the large methods and make it a bit more readable, but is fine for now and looks good otherwise.

File: data-access/mandatees-decisions.ts (hunk 13 above)
Comment (JonasVanHoof) [reply data-access/mandatees-decisions.ts]: Done in last commit 👌🏻

**Pull request merged**